File: go/internal/profiles/loader.go
package profiles

import (
	"fmt"
	"log/slog"
	"path/filepath"
	"sort"
	"strings"

	"github.com/chadit/LinodeMCP/internal/config"
)

// wildcardChar is the only glob metacharacter recognized in profile tool
// patterns. Anything else in a pattern is treated literally. Keeping this as
// a constant documents the intentional limit (no regex, no character
// classes).
const wildcardChar = "*"

// ResolveActiveProfile resolves the configured active profile against the
// supplied tool registry and returns the resulting Profile.
//
// Lookup order:
//
//  1. If cfg.ActiveProfile is empty, treat it as the built-in "default".
//  2. If the name matches a built-in, honor any
//     cfg.ProfilesBuiltinOverrides[name].Disabled and return the resolved
//     built-in Profile.
//  3. If the name matches a key in cfg.Profiles, build a Profile by
//     expanding wildcards in AllowedTools and DeniedTools against the
//     registry (denied wins over allowed).
//  4. Otherwise return ErrActiveProfileUnknown.
//
// The registry slice carries (tool name, capability) pairs. Callers (the
// server in Phase 4, tests today) build it from their own tool inventory and
// pass it in. Keeping the registry an argument rather than a package-level
// import keeps internal/profiles free of cycles with internal/server and
// internal/tools.
func ResolveActiveProfile(cfg *config.Config, registry []ToolDescriptor) (Profile, error) {
	if cfg == nil {
		return Profile{}, fmt.Errorf("active profile %q: %w", BuiltinDefault, ErrActiveProfileUnknown)
	}

	name := cfg.ActiveProfile
	if name == "" {
		name = BuiltinDefault
	}

	builtins := BuiltinProfiles(registry)

	if builtin, ok := builtins[name]; ok {
		builtinCopy := builtin

		return resolveBuiltin(name, &builtinCopy, cfg)
	}

	if userCfg, ok := cfg.Profiles[name]; ok {
		warnOverrideIgnored(name, cfg)

		userCopy := userCfg

		return resolveUserDefined(name, &userCopy, registry), nil
	}

	return Profile{}, fmt.Errorf("active profile %q: %w", name, ErrActiveProfileUnknown)
}

// LookupProfile resolves a profile by name across both built-ins and
// user-defined entries, returning the materialized Profile and true on
// hit, or the zero Profile and false on miss. Unlike ResolveActiveProfile
// it ignores the Disabled flag: the Phase 8 builder uses this to clone
// from any profile in the catalog, including disabled built-ins like
// full-access and emergency. User-defined entries shadow built-ins by
// name, matching the precedence ResolveActiveProfile uses.
//
// nil cfg or empty name both return (Profile{}, false); callers that
// need empty-equals-default semantics should fall back themselves.
func LookupProfile(name string, cfg *config.Config, registry []ToolDescriptor) (Profile, bool) {
	if cfg == nil || name == "" {
		return Profile{}, false
	}

	if userCfg, ok := cfg.Profiles[name]; ok {
		userCopy := userCfg

		return resolveUserDefined(name, &userCopy, registry), true
	}

	builtins := BuiltinProfiles(registry)
	if builtin, ok := builtins[name]; ok {
		// Strip Disabled so a cloned-from-disabled draft doesn't carry
		// the flag into the new profile. The user can re-disable via
		// config after the save.
		resolved := builtin
		resolved.Disabled = false

		return resolved, true
	}

	return Profile{}, false
}

// resolveBuiltin applies override toggles (currently just Disabled) to a
// resolved built-in profile. The returned Profile keeps the catalog's
// AllowedTools verbatim; only the disable state is mutable from config.
func resolveBuiltin(name string, builtin *Profile, cfg *config.Config) (Profile, error) {
	override, hasOverride := cfg.ProfilesBuiltinOverrides[name]

	disabled := builtin.Disabled
	if hasOverride {
		disabled = override.Disabled
	}

	if disabled {
		return Profile{}, fmt.Errorf("active profile %q: %w", name, ErrActiveProfileDisabled)
	}

	resolved := *builtin
	resolved.Disabled = false

	return resolved, nil
}

// resolveUserDefined materializes a Profile from a UserProfileConfig by
// expanding wildcards in AllowedTools and DeniedTools against the registry.
// Denied entries are subtracted from the allowed set after expansion so an
// explicit deny always wins over a wildcard allow.
//
// Patterns that match no tool in the registry, and literal names that do not
// appear in the registry, produce a warning but never an error: the spec
// treats both as drift warnings the user can ignore or fix.
func resolveUserDefined(name string, userCfg *config.UserProfileConfig, registry []ToolDescriptor) Profile {
	allowed := expandPatterns(name, "allowed_tools", userCfg.AllowedTools, registry)
	denied := expandPatterns(name, "denied_tools", userCfg.DeniedTools, registry)

	for tool := range denied {
		delete(allowed, tool)
	}

	tools := make([]string, 0, len(allowed))
	for tool := range allowed {
		tools = append(tools, tool)
	}

	sort.Strings(tools)

	return Profile{
		Name:                name,
		Description:         userCfg.Description,
		AllowedTools:        tools,
		AllowedEnvironments: append([]string(nil), userCfg.AllowedEnvironments...),
		RequiredTokenScopes: append([]string(nil), userCfg.RequiredTokenScopes...),
		AllowYolo:           userCfg.AllowYolo,
		Disabled:            false,
	}
}

// expandPatterns walks the supplied patterns and returns the set of registry
// tool names they cover. Wildcards (entries containing "*") are matched with
// shell-glob semantics via filepath.Match; literal entries must match an
// existing tool name. Both forms emit a warning when they resolve to nothing.
func expandPatterns(profileName, fieldName string, patterns []string, registry []ToolDescriptor) map[string]struct{} {
	result := make(map[string]struct{}, len(registry))

	for _, pattern := range patterns {
		if pattern == "" {
			continue
		}

		matches := matchPattern(pattern, registry)
		if len(matches) == 0 {
			logUnmatchedPattern(profileName, fieldName, pattern)

			continue
		}

		for _, name := range matches {
			result[name] = struct{}{}
		}
	}

	return result
}

// matchPattern returns the registry tool names that match the given pattern.
// If the pattern contains "*", filepath.Match is used; otherwise the pattern
// must equal a registered tool name exactly. Malformed glob patterns (e.g. an
// unmatched bracket) match nothing, which surfaces as an unmatched-pattern
// warning to the caller.
func matchPattern(pattern string, registry []ToolDescriptor) []string {
	matches := make([]string, 0)

	if !strings.Contains(pattern, wildcardChar) {
		for _, descriptor := range registry {
			if descriptor.Name == pattern {
				matches = append(matches, descriptor.Name)

				break
			}
		}

		return matches
	}

	for _, descriptor := range registry {
		ok, err := filepath.Match(pattern, descriptor.Name)
		if err != nil || !ok {
			continue
		}

		matches = append(matches, descriptor.Name)
	}

	return matches
}

// logUnmatchedPattern emits a warning for a pattern that resolved to zero
// tools. Literal misses and wildcard misses use the same log line shape so
// operators see consistent diagnostics regardless of which form they used.
func logUnmatchedPattern(profileName, fieldName, pattern string) {
	slog.Warn(
		"profile pattern matched no registered tools",
		"profile", profileName,
		"field", fieldName,
		"pattern", pattern,
	)
}

// warnOverrideIgnored emits a warning when a user-defined profile name
// appears in ProfilesBuiltinOverrides. Overrides only apply to built-ins; the
// entry is ignored for user-defined profiles. The check is best-effort and
// fires only for the active profile to avoid noisy startup logs.
func warnOverrideIgnored(name string, cfg *config.Config) {
	if _, ok := cfg.ProfilesBuiltinOverrides[name]; !ok {
		return
	}

	slog.Warn(
		"builtin override ignored for user-defined profile",
		"profile", name,
	)
}

File: python/src/linodemcp/profiles/loader.py
"""Active-profile resolution against the live tool registry.

Pure function with no I/O: takes a parsed ``Config`` plus the registered
tool descriptors and returns a ``Profile`` whose ``allowed_tools`` is the
explicit, expanded list the registration filter will iterate at Phase 4.

The module deliberately does NOT import ``linodemcp.server`` or
``linodemcp.tools``. Phase 4 will wire the resolver into the server at
startup; this module's job ends with returning the resolved ``Profile``.

Wildcard semantics use ``fnmatch.fnmatch`` (shell-glob: only ``*`` is
special). Resolution order: expand ``allowed_tools`` against the registry,
then subtract expanded ``denied_tools``. Built-in profile overrides
(``profiles_builtin_overrides[name].disabled``) refuse a disabled built-in;
overrides naming a non-built-in are logged and ignored.
"""

from __future__ import annotations

import fnmatch
import logging
from typing import TYPE_CHECKING

from linodemcp.profiles.builtin import builtin_profiles, has_mutating_tools
from linodemcp.profiles.errors import (
    ActiveProfileDisabledError,
    ActiveProfileUnknownError,
)
from linodemcp.profiles.profile import Profile

if TYPE_CHECKING:
    from collections.abc import Sequence

    from linodemcp.config import Config, UserProfileConfig
    from linodemcp.profiles.builtin import ToolDescriptor

logger = logging.getLogger(__name__)

DEFAULT_PROFILE_NAME = "default"

__all__ = ["DEFAULT_PROFILE_NAME", "lookup_profile", "resolve_active_profile"]


def lookup_profile(
    name: str,
    cfg: Config,
    registry: Sequence[ToolDescriptor],
) -> Profile | None:
    """Resolve a profile by name across built-ins and user-defined entries.

    Mirror of Go ``profiles.LookupProfile``. The Phase 8.3 ``_draft_new``
    handler uses this to seed a new draft from the named ``clone_from``
    profile. Unlike :func:`resolve_active_profile` this ignores the
    Disabled flag: callers may clone from a disabled built-in like
    ``full-access`` or ``emergency``. User-defined entries shadow
    built-ins by name (same precedence as the active resolver).

    Returns the materialized ``Profile`` on hit or ``None`` on miss.
    Empty ``name`` returns ``None``; callers that need
    empty-equals-default semantics fall back themselves.
    """
    if not name:
        return None

    if name in cfg.profiles:
        return _resolve_user_profile(name, cfg.profiles[name], registry)

    builtins = builtin_profiles(registry)
    found = builtins.get(name)
    if found is None:
        return None

    # Strip the disabled flag so a clone from a disabled built-in
    # doesn't carry the flag into the new draft.
    return Profile(
        name=found.name,
        description=found.description,
        allowed_tools=found.allowed_tools,
        allowed_environments=found.allowed_environments,
        required_token_scopes=found.required_token_scopes,
        elevated=found.elevated,
        allow_yolo=found.allow_yolo,
        disabled=False,
    )


def _expand_patterns(
    patterns: Sequence[str],
    registry_names: Sequence[str],
    field_label: str,
) -> set[str]:
    """Expand a list of literal or wildcard patterns against the registry.

    Each pattern is fed through ``fnmatch.fnmatch`` so callers get shell-glob
    semantics. A pattern containing ``*`` that matches nothing logs a
    warning. A literal name not present in the registry also logs a warning.
    Both cases drop silently from the resolved set (load continues).

    ``field_label`` shows up in the warning so users can tell whether the
    bad entry was in ``allowed_tools`` or ``denied_tools``.
    """
    registry_set = set(registry_names)
    resolved: set[str] = set()
    for pattern in patterns:
        if "*" in pattern:
            matches = {
                name for name in registry_names if fnmatch.fnmatch(name, pattern)
            }
            if not matches:
                logger.warning(
                    "profile %s pattern %r matched no registered tools",
                    field_label,
                    pattern,
                )
                continue
            resolved.update(matches)
            continue
        if pattern in registry_set:
            resolved.add(pattern)
            continue
        logger.warning(
            "profile %s entry %r is not a registered tool name",
            field_label,
            pattern,
        )
    return resolved


def _resolve_user_profile(
    name: str,
    entry: UserProfileConfig,
    registry: Sequence[ToolDescriptor],
) -> Profile:
    """Build a ``Profile`` from a user-defined config entry.

    Order of operations matches the spec: expand ``allowed_tools``, then
    subtract expanded ``denied_tools``. Explicit deny wins over allow.
    Tool names are sorted in the returned tuple so the resolved profile is
    stable for log output, parity comparison, and golden-file tests.
    """
    registry_names = [tool.name for tool in registry]
    allowed = _expand_patterns(entry.allowed_tools, registry_names, "allowed_tools")
    denied = _expand_patterns(entry.denied_tools, registry_names, "denied_tools")
    resolved_names = tuple(sorted(allowed - denied))
    return Profile(
        name=name,
        description=entry.description,
        allowed_tools=resolved_names,
        allowed_environments=entry.allowed_environments,
        required_token_scopes=entry.required_token_scopes,
        # Derived from the resolved tool list, never from user config:
        # the missing-token policy must reflect what the profile can
        # actually mutate.
        elevated=has_mutating_tools(registry, resolved_names),
        allow_yolo=entry.allow_yolo,
        disabled=False,
    )


def resolve_active_profile(
    cfg: Config,
    registry: Sequence[ToolDescriptor],
) -> Profile:
    """Resolve ``cfg.active_profile`` against the built-in catalog and user entries.

    Falls back to the ``default`` built-in when ``cfg.active_profile`` is
    empty. Returns the resolved ``Profile`` whose ``allowed_tools`` is the
    explicit list the registration filter consumes in Phase 4.

    Raises:
        ActiveProfileDisabledError: the active name is a built-in disabled
            via ``profiles_builtin_overrides``.
        ActiveProfileUnknownError: the active name is neither a built-in
            nor a user-defined entry.
    """
    requested = cfg.active_profile.strip() or DEFAULT_PROFILE_NAME
    builtins = builtin_profiles(registry)

    # Surface overrides aimed at non-built-in names so users notice the
    # typo: only built-ins can be disabled, so an override on a user-defined
    # name silently does nothing.
    for override_name in cfg.profiles_builtin_overrides:
        if override_name not in builtins:
            logger.warning(
                "profiles_builtin_overrides entry %r does not name a built-in "
                "profile; override ignored",
                override_name,
            )

    if requested in builtins:
        override = cfg.profiles_builtin_overrides.get(requested)
        if override is not None and override.disabled:
            msg = (
                f"active profile {requested!r} is a built-in disabled via "
                "profiles_builtin_overrides; enable it in config or pick "
                "a different profile"
            )
            raise ActiveProfileDisabledError(msg)
        return builtins[requested]

    if requested in cfg.profiles:
        return _resolve_user_profile(requested, cfg.profiles[requested], registry)

    msg = (
        f"active profile {requested!r} is not a known built-in and is not "
        "defined under 'profiles:' in config"
    )
    raise ActiveProfileUnknownError(msg)
